pkg/sql/catalog/descriptor.go

```diff
@@ -632,11 +632,21 @@ type TableDescriptor interface {
 	// It's only non-nil if IsView is true.
 	GetDependsOnTypes() []descpb.ID
 
-	// GetConstraintInfoWithLookup returns a summary of all constraints on the
+	// GetNonDropConstraintInfoWithLookup returns a summary of all constraints on the
 	// table using the provided function to fetch a TableDescriptor from an ID.
-	GetConstraintInfoWithLookup(fn TableLookupFn) (map[string]descpb.ConstraintDetail, error)
-	// GetConstraintInfo returns a summary of all constraints on the table.
-	GetConstraintInfo() (map[string]descpb.ConstraintDetail, error)
+	GetNonDropConstraintInfoWithLookup(fn TableLookupFn) (map[string]descpb.ConstraintDetail, error)
+	// GetAllConstraintInfoWithLookup is similar to GetActiveAndInactiveConstraintInfoWithLookup
+	// except that
+	//  - it also includes constraints that are being dropped;
+	//  - the return uses `descpb.ConstraintID` as unique identifier.
+	GetAllConstraintInfoWithLookup(fn TableLookupFn) (map[descpb.ConstraintID]descpb.ConstraintDetail, error)
+	// GetNonDropConstraintInfo returns a summary of all constraints on the table.
+	GetNonDropConstraintInfo() (map[string]descpb.ConstraintDetail, error)
+	// GetAllConstraintInfo is similar to GetActiveAndInactiveConstraint
+	// except that
+	//  - it also includes constraints that are being dropped;
+	//  - the return uses `descpb.ConstraintID` as unique identifier.
+	GetAllConstraintInfo() (map[descpb.ConstraintID]descpb.ConstraintDetail, error)
```

Comment:
The existing `GetConstraintInfo*` methods have been carried over from a long time ago and have remained largely unchanged. We should improve the `TableDescriptor` interface by removing them and replacing them with more useful methods, like the constraint-slice methods you're adding in this PR, rather than doubling down on something undesirable. Personally I particularly don't like the fact that they return maps, and that they require so much code _inside_ the table descriptor interface implementation.

You might be able to simplify things by moving all the `collectAllConstraintInfo` code you copy-pasted into a function (or two) in `catalog` which takes a lookup fn and calls your new methods to build a map, and use that function to get rid of `GetConstraintInfo*` entirely.